crates/compiler/gen_dev/src/generic64/mod.rs

```diff
@@ -5367,6 +5377,46 @@ impl<
     }
 }
 
+fn load_union_field_ptr_at_index_help<'a>(
+    layout_interner: &'a STLayoutInterner<'a>,
+    indices: &[u64],
+    layouts: &[InLayout<'a>],
+    mut offset: u32,
+) -> u32 {
+    debug_assert_ne!(indices.len(), 0);
+
+    let idx = indices[0];
+    for field in &layouts[..idx as usize] {
+        offset += layout_interner.stack_size(*field);
+    }
+
+    if indices.len() <= 1 {
+        return offset;
+    }
+    use LayoutRepr::*;
+    let enclosing_layout = layouts[idx as usize];
+    return match layout_interner.get_repr(enclosing_layout) {
+        Struct(layouts) => {
+            load_union_field_ptr_at_index_help(layout_interner, &indices[1..], layouts, offset)
+        },
+        Union(UnionLayout::NonRecursive(tags)) => {
+            // apart from the current index, to get information out of a
+            // non-recursive tag union we need 2 more number: a tag id and an index
+            debug_assert!(indices.len() >= 3);
+            let tag_id = indices[1] as usize;
+            let tag_layout = tags[tag_id];
+
+            load_union_field_ptr_at_index_help(layout_interner, &indices[2..], tag_layout, offset)
+        }
+        //TODO: can any of these be a valid target for getting a pointer out of their inner layouts?
+        //(so not them being pointers)
+        Union(_) | Builtin(_) | Ptr(_) | RecursivePointer(_) | LambdaSet(_) | FunctionPointer(_)
+        | Erased(_) => unreachable!(
+            "Following the path of the indices, a layout has been reached where the address of the desired pointer is behind another pointer"
+        ),
```

Comment:
lambda set probably could (it should just recurse on its inner type). Also the error message is not correct for `Builtin`. The remainder (other union values,  ptr, recursiveptr, functionpointer and erased) are all pointer types. I'd say just give builtin its own branch with a custom error message.

Reply:
For the lambda set case, I believe the access path should be at most 1 right? Since I believe we will never explicitly TR-optimize a lambda set. If so, let's add a debug assertion for the size of the `indices` in that case.